src/lib.rs

```diff
@@ -2,269 +2,186 @@
+    })
+}
 
-    #[test]
-    fn uncompress_lzip() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "lzip"]).unwrap();
-        uncompress("test.lzip", &dir_handle.path().join("target"), Kind::LZip)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.lzip", "lzip -dc"]);
-    }
+// reads the input file headers and writes them to disk.
+pub fn uncompress(
+    source_path: String,
+    target_path: String,
+    block_size: usize,
+) -> compress::Result<()> {
+    init_archive(|mut entry, ar, aw| unsafe {
+        if archive_read_open_filename(
+            ar,
+            CString::new(source_path.clone())?.into_raw(),
+            block_size,
+        ) == ARCHIVE_OK
+        {
+            loop {
+                match archive_read_next_header(ar, &mut entry) {
+                    ARCHIVE_EOF => return Ok(()),
+                    ARCHIVE_WARN | ARCHIVE_FAILED | ARCHIVE_FATAL => {
+                        return Err(compress::Error::from(archive_error_string(ar)))
+                    }
+                    ARCHIVE_OK => {
+                        let file_name = archive_entry_pathname(entry);
+
+                        if file_name.is_null() {
+                            return Err(compress::Error::EntryNull);
+                        }
+
+                        let mut target_path = target_path.clone();
+                        if target_path.ends_with('/') {
+                            target_path.pop();
+                        }
+
+                        archive_entry_set_pathname(entry, format_path(target_path, file_name)?);
+
+                        handle_error(archive_write_header(aw, entry), aw)?;
+                        copy_data(ar, aw)?;
+                        handle_error(archive_write_finish_entry(aw), aw)?;
+                    }
+                    _ => unreachable!("Error, the impossible has happened."),
+                }
+            }
+        }
+        Ok(())
+    })
+}
 
-    #[test]
-    fn uncompress_zip() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "unzip"]).unwrap();
-        uncompress("test.zip", &dir_handle.path().join("target"), Kind::Zip)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.zip", "unzip - -o -d"]);
-    }
+pub fn get_file_from_archive(
+    source_path: String,
+    target_file: String,
+    target_path: String,
+    block_size: usize,
+) -> compress::Result<()> {
+    init_archive(|mut entry, ar, aw| unsafe {
+        if archive_read_open_filename(
+            ar,
+            CString::new(source_path.clone())?.into_raw(),
+            block_size,
+        ) == ARCHIVE_OK
+        {
+            loop {
+                match archive_read_next_header(ar, &mut entry) {
+                    ARCHIVE_EOF => {
+                        return Err(compress::Error::ExtractionError(String::from(
+                            "File not found.",
+                        )))
```

Comment:
I think this would be a fairly common error to happen, I think its better to have a specific enum member for it.